dns_server/src/dns_server/mod.rs

```diff
@@ -0,0 +1,287 @@
+            );
+        }
+
+        if let Some(nameserver) = self.nameserver.as_ref() {
+            let mut ns_rec = RecordSet::with_ttl(self.host.clone(), RecordType::NS, TTL_NS);
+            ns_rec.add_rdata(RData::NS(nameserver.clone()));
+            new_records.insert(
+                RrKey::new(ns_rec.name().clone().into(), ns_rec.record_type()),
+                Arc::new(ns_rec),
+            );
+        }
+
+        // A records
+        let mut ipv4_rec = RecordSet::with_ttl(self.host.clone(), RecordType::A, TTL_IP);
+        for ip in ipv4 {
+            ipv4_rec.add_rdata(RData::A(ip));
+        }
+        new_records.insert(
+            RrKey::new(ipv4_rec.name().clone().into(), ipv4_rec.record_type()),
+            Arc::new(ipv4_rec),
+        );
+
+        // AAAA records
+        let mut ipv6_rec = RecordSet::with_ttl(self.host.clone(), RecordType::AAAA, TTL_IP);
+        for ip in ipv6 {
+            ipv6_rec.add_rdata(RData::AAAA(ip));
+        }
+        new_records.insert(
+            RrKey::new(ipv6_rec.name().clone().into(), ipv6_rec.record_type()),
+            Arc::new(ipv6_rec),
+        );
+
+        *self.inner.records_mut().await = new_records;
+    }
+}
+
+#[async_trait::async_trait]
+impl Authority for AuthorityImpl {
+    type Lookup = AuthLookup;
+
+    fn zone_type(&self) -> ZoneType {
+        self.inner.zone_type()
+    }
+
+    fn is_axfr_allowed(&self) -> bool {
+        self.inner.is_axfr_allowed()
+    }
+
+    async fn update(&self, update: &MessageRequest) -> UpdateResult<bool> {
+        self.inner.update(update).await
+    }
+
+    fn origin(&self) -> &LowerName {
+        self.inner.origin()
+    }
+
+    async fn lookup(
+        &self,
+        name: &LowerName,
+        query_type: RecordType,
+        lookup_options: LookupOptions,
+    ) -> Result<Self::Lookup, LookupError> {
+        self.refresh().await;
+        self.inner.lookup(name, query_type, lookup_options).await
+    }
+
+    async fn search(
+        &self,
+        request_info: RequestInfo<'_>,
+        lookup_options: LookupOptions,
+    ) -> Result<Self::Lookup, LookupError> {
+        self.refresh().await;
+        self.inner.search(request_info, lookup_options).await
+    }
+
+    async fn get_nsec_records(
+        &self,
+        name: &LowerName,
+        lookup_options: LookupOptions,
+    ) -> Result<Self::Lookup, LookupError> {
+        self.inner.get_nsec_records(name, lookup_options).await
+    }
+}
+
+fn handle_command(auth: &AuthorityImpl, command: ServerCommands) {
+    match command {
+        ServerCommands::AddAddress(IpAddr::V4(ip)) => {
+            auth.ip4.lock().expect("mutex must be valid (ipv4)").push(ip);
+        }
+        ServerCommands::AddAddress(IpAddr::V6(ip)) => {
+            auth.ip6.lock().expect("mutex must be valid (ipv6)").push(ip);
+        }
+        ServerCommands::DelAddress(IpAddr::V4(ip)) => {
+            auth.ip4.lock().expect("mutex must be valid (ipv4)").retain(|val| *val != ip);
```

Comment:
Again, I see "mutex must be valid (ipv4)" repeated twice. Please make the repeated expect messages distinguishable.

Reply:
Updated. I did not pay much attention because `expect` would also print the filename and line number, which is usually enough.

Reply:
That assumes you know what version you're testing against. It often happens that we don't know what version caused an issue with a user.